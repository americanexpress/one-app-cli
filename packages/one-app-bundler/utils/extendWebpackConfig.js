/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const path = require('path');
const merge = require('webpack-merge');
const getConfigOptions = require('./getConfigOptions');
const getCliOptions = require('./getCliOptions');
const createResolver = require('../webpack/createResolver');

function getCustomWebpackConfigPath(options, bundleTarget) {
  const { webpackConfigPath, webpackClientConfigPath, webpackServerConfigPath } = options;

  if (webpackConfigPath) return webpackConfigPath;
  if (bundleTarget === 'client' && webpackClientConfigPath) return webpackClientConfigPath;
  if (bundleTarget === 'server' && webpackServerConfigPath) return webpackServerConfigPath;

  return false;
}

function extendWebpackConfig(webpackConfig, bundleTarget) {
  const configOptions = getConfigOptions();
  const cliOptions = getCliOptions();
  const { mainFields } = webpackConfig.resolve;
  const resolve = createResolver({ mainFields, resolveToContext: true });
  const {
    appCompatibility,
    requiredExternals,
    providedExternals,
  } = configOptions;
  const { watch } = cliOptions;

  const customWebpackConfigPath = getCustomWebpackConfigPath(configOptions, bundleTarget);
  let customWebpackConfig = {};

  if (customWebpackConfigPath) {
    // Dynamic require is needed here for loading custom config
    // eslint-disable-next-line global-require, import/no-dynamic-require
    customWebpackConfig = require(path.join(process.cwd(), customWebpackConfigPath));
  }

  const indexPath = path.join(process.cwd(), 'src', 'index');

  if (providedExternals) {
    customWebpackConfig = merge(customWebpackConfig, {
      module: {
        rules: [{
          test: indexPath,
          use: [{
            loader: '@americanexpress/one-app-bundler/webpack/loaders/provided-externals-loader',
            options: {
              providedExternals,
            },
          }],
        }],
      },
    });
  }

  if (requiredExternals) {
    customWebpackConfig = merge(customWebpackConfig, {
      module: {
        rules: [...requiredExternals.map((externalName) => ({
          test: resolve(externalName),
          use: [{
            loader: '@americanexpress/one-app-bundler/webpack/loaders/externals-loader',
            options: {
              externalName,
            },
          }],
        })), {
          test: indexPath,
          use: [{
            loader: '@americanexpress/one-app-bundler/webpack/loaders/validate-required-externals-loader',
            options: {
              requiredExternals,
            },
          }],
        }],
      },
    });
  }

  if (appCompatibility) {
    customWebpackConfig = merge(customWebpackConfig, {
      module: {
        rules: [{
          test: indexPath,
          use: [{
            loader: '@americanexpress/one-app-bundler/webpack/loaders/validate-one-app-compatibility-loader',
            options: {
              appCompatibility,
            },
          }],
        }],
      },
    });
  }

  if (watch) {
    customWebpackConfig = merge(customWebpackConfig, {
      watch: true,
      watchOptions: {
        aggregateTimeout: 500,
        ignored: /node_modules/,
      },
    });
  }

  return merge(webpackConfig, customWebpackConfig);
}

module.exports = extendWebpackConfig;
