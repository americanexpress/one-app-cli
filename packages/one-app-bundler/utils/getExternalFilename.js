/**
 * Returns the external's filename
 * @param {string} externalName the external name
 * @returns {string}
 */
const getExternalFilename = (externalName) => `${externalName}.js`;

module.exports = getExternalFilename;
