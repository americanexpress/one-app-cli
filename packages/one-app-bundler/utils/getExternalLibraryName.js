const { snakeCase } = require('lodash');
const { EXTERNAL_PREFIX } = require('..');

/**
 * Creates a unique global variable name to store an external's code
 * @param {string} name External name
 * @param {string} version External semver version (e.g. 1.2.3)
 * @returns {string} global variable name
 */
const getExternalLibraryName = (name, version) => [EXTERNAL_PREFIX, snakeCase(name), version.replace(/[^\d.]+/g, '').replace(/\.+/g, '_')].filter(Boolean).join('__');

module.exports = getExternalLibraryName;
