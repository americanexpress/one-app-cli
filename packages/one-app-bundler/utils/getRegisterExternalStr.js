const getExternalLibraryName = require('./getExternalLibraryName');

const getRegisterExternalStr = (name, version) => `Holocron.registerExternal({ name: "${name}", version: "${version}", module: ${getExternalLibraryName(name)}});`;

module.exports = getRegisterExternalStr;
