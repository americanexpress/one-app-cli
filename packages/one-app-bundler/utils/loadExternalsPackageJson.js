/*
 * Copyright 2023 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */

// Because this file uses very modern node features, it may not be parsed by every IDE.
// As long as eslint and jest are happy when they are run by yarn, all is ok.
const loadExternalsPackageJson = async (externalName) => (await import(`${externalName}/package.json`, { assert: { type: "json" } })).default;

export default loadExternalsPackageJson;
