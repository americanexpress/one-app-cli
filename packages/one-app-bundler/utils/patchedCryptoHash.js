const crypto = require('crypto');

// Monkey Patch for unsupported hash algo. Needed to support Node >=17.
// https://github.com/webpack/webpack/issues/13572#issuecomment-923736472
const originalCreateHash = crypto.createHash;
crypto.createHash = (algo) => originalCreateHash(algo === 'md4' ? 'sha256' : algo);
