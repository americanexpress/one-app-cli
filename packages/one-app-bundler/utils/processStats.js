/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

const fs = require('fs');
const path = require('path');
const chalk = require('chalk');

const generateIntegrityManifest = require('./generateIntegrityManifest');

module.exports = function processStats(stats, watch) {
  const jsonStats = stats.toJson();

  if (stats.hasErrors()) {
    const errorMessages = jsonStats.errors.map((e) => ([chalk.red(e), chalk.red(e.stack)].join('\n')).join('\n'));
    console.error(chalk.red(`\nerror - "${stats.compilation.name}":\n`), errorMessages);
    process.exitCode = 1;
    return;
  }

  if (stats.hasWarnings()) {
    const warningMessages = jsonStats.warnings.join('\n');
    console.warn(chalk.yellow(`\nwarning - "${stats.compilation.name}":\n`), warningMessages);
  }

  fs.writeFileSync(path.join(process.cwd(), `.webpack-stats.${stats.compilation.name}.json`), JSON.stringify(jsonStats));

  generateIntegrityManifest(
    stats.compilation.name,
    path.join(stats.compilation.compiler.outputPath, stats.compilation.outputOptions.filename)
  );

  if (watch) {
    return;
  }

  // for CI/CD have webpack exit if a module isn't found (should be default behavior but isn't)
  if (stats.compilation.errors[0]) {
    throw stats.compilation.errors[0];
  }
};
