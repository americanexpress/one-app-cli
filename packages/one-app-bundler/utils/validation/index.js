/*
 * Copyright 2020 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing
 * permissions and limitations under the License.
 */
const Joi = require('@hapi/joi');

const externalsSchema = Joi.array().items(Joi.string().required()).messages({
  'array.base': 'Externals must be an array',
  'array.includesRequiredUnknowns': 'Externals must have at least one entry',
  'string.base': 'Externals must contain strings',
});

const webpackConfigSchema = Joi.string().messages({
  'string.base': 'Webpack Configs must be a string',
});

const performanceBudgetSchema = Joi.number().strict();

const appSchema = Joi.object().keys({
  compatibility: Joi.string(),
});

const purgecssSchema = Joi.object({
  paths: Joi.array().items(Joi.string().required()),
  extractors: Joi.array().items(Joi.object().keys({
    extractor: Joi.string(),
    extensions: Joi.array().items(Joi.string().required()),
  }).required()),
  fontFace: Joi.boolean().strict(),
  keyframes: Joi.boolean().strict(),
  variables: Joi.boolean().strict(),
  whitelist: Joi.array().items(Joi.string().required()),
  whitelistPatterns: Joi.array().items(Joi.string().required()),
  whitelistPatternsChildren: Joi.array().items(Joi.string().required()),
  disabled: Joi.boolean().strict(),
});

const optionsSchema = Joi.object({
  providedExternals: externalsSchema,
  requiredExternals: externalsSchema,
  performanceBudget: performanceBudgetSchema,
  app: appSchema,
  purgecss: purgecssSchema,
  webpackConfigPath: webpackConfigSchema,
  webpackClientConfigPath: webpackConfigSchema,
  webpackServerConfigPath: webpackConfigSchema,
});

function validateSchema(schema, validationTarget) {
  const { error, value } = schema.validate(
    validationTarget,
    { abortEarly: false }
  );
  if (error) throw error;

  return value;
}

function validateBundler(options) {
  return validateSchema(optionsSchema, options);
}

module.exports = {
  externalsSchema,
  webpackConfigSchema,
  performanceBudgetSchema,
  appSchema,
  purgecssSchema,
  validateBundler,
};
