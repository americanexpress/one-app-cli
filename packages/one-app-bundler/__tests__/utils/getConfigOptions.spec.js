/*
 * Copyright 2019 American Express Travel Related Services Company, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License
 * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
 * or implied. See the License for the specific language governing permissions and limitations
 * under the License.
 */

jest.mock('read-pkg-up', () => ({ sync: jest.fn(() => ({ pkg: {} })) }));

describe('getConfigOptions', () => {
  let readPkgUp;

  const consoleWarnSpy = jest.spyOn(console, 'warn');

  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetModules();
    readPkgUp = require('read-pkg-up');
  });

  it('should handle missing one-amex config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: {} });
    const getConfigOptions = require('../../utils/getConfigOptions');
    expect(getConfigOptions).not.toThrow();
  });

  it('should handle missing bundler config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { runner: {} } } });
    const getConfigOptions = require('../../utils/getConfigOptions');
    expect(getConfigOptions).not.toThrow();
  });

  it('should include app compatability', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { app: { compatibility: '^5.1.0' } } } });
    const getConfigOptions = require('../../utils/getConfigOptions');
    expect(getConfigOptions()).toMatchObject({ appCompatibility: '^5.1.0' });
  });

  it('should default purgecss to an empty object', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: {} });
    const getConfigOptions = require('../../utils/getConfigOptions');
    expect(getConfigOptions()).toMatchObject({ purgecss: {} });
  });

  it('should not override an existing purgecss config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { purgecss: { paths: ['foo'] } } } } });
    const getConfigOptions = require('../../utils/getConfigOptions');
    expect(getConfigOptions()).toMatchObject({ purgecss: { paths: ['foo'] } });
  });

  it('should throw when a user includes both requiredExternals and providedExternals configs', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { requiredExternals: ['a'], providedExternals: ['b'] } } } });
    expect(() => require('../../utils/getConfigOptions')).toThrowErrorMatchingSnapshot();
  });

  it('should allow a user to include requiredExternals', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { providedExternals: ['b'] } } } });
    expect(() => require('../../utils/getConfigOptions')).not.toThrow();
  });

  it('should throw when a user attempts to provide an app provided external', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { providedExternals: ['@americanexpress/one-app-ducks'] } } } });
    expect(() => require('../../utils/getConfigOptions')).toThrowErrorMatchingSnapshot();
  });

  it('should throw when a user attempts to use an app provided external from the root module', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { requiredExternals: ['@americanexpress/one-app-router'] } } } });
    expect(() => require('../../utils/getConfigOptions')).toThrowErrorMatchingSnapshot();
  });

  it('should warn when the user provides a custom webpack config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { webpackConfigPath: 'webpack.config.js' } } } });
    require('../../utils/getConfigOptions');
    expect(consoleWarnSpy).toHaveBeenCalledWith('@americanexpress/one-app-bundler: Using a custom webpack config can cause unintended side effects. Issues resulting from custom configuration will not be supported.');
  });

  it('should warn when the user provides a custom client webpack config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { webpackClientConfigPath: 'webpack.client.config.js' } } } });
    require('../../utils/getConfigOptions');
    expect(consoleWarnSpy).toHaveBeenCalledWith('@americanexpress/one-app-bundler: Using a custom webpack config can cause unintended side effects. Issues resulting from custom configuration will not be supported.');
  });

  it('should warn when the user provides a custom server webpack config', () => {
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { webpackServerConfigPath: 'webpack.config.js' } } } });
    require('../../utils/getConfigOptions');
    expect(consoleWarnSpy).toHaveBeenCalledWith('@americanexpress/one-app-bundler: Using a custom webpack config can cause unintended side effects. Issues resulting from custom configuration will not be supported.');
  });

  it('should throw when a user attempts to use both webpackConfigPath and webpackClientConfigPath', () => {
    const errorRegex = /@americanexpress\/one-app-bundler: Modules cannot configure both webpackConfigPath and webpackClientConfigPath or webpackServerConfigPath. See README for details./;
    readPkgUp.sync.mockReturnValueOnce({ packageJson: { 'one-amex': { bundler: { webpackConfigPath: 'webpack.config.js', webpackClientConfigPath: 'webpack.config.js' } } } });
    expect(() => require('../../utils/getConfigOptions')).toThrow(errorRegex);
  });
});
