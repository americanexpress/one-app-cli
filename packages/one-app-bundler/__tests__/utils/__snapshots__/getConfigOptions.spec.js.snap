// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getConfigOptions should throw when a user attempts to provide an app provided external 1`] = `"@americanexpress/one-app-bundler: Attempted to bundle @americanexpress/one-app-ducks, but modules cannot provide externals that One App includes."`;

exports[`getConfigOptions should throw when a user attempts to use an app provided external from the root module 1`] = `"@americanexpress/one-app-bundler: Attempted to bundle @americanexpress/one-app-router, but modules cannot provide externals that One App includes."`;

exports[`getConfigOptions should throw when a user includes both requiredExternals and providedExternals configs 1`] = `"@americanexpress/one-app-bundler: Modules cannot configure both requiredExternals and providedExternals. See README for details."`;
